internal/credentials/keyring.go

```diff
@@ -3,44 +3,177 @@
 package credentials
 
 import (
-	"strings"
+	"encoding/json"
+	"fmt"
 
+	"github.com/google/uuid"
+	"github.com/posit-dev/publisher/internal/logging"
+	"github.com/posit-dev/publisher/internal/util"
 	"github.com/zalando/go-keyring"
 )
 
-type KeyringService struct {
-	keyring.Keyring
+type keyringCredentialsService struct {
+	log logging.Logger
 }
 
-func (ks *KeyringService) Set(service, user, password string) error {
-	err := keyring.Set(service, user, password)
+func NewKeyringCredentialsService(log logging.Logger) *keyringCredentialsService {
+	return &keyringCredentialsService{
+		log: log,
+	}
+}
+
+func (ks *keyringCredentialsService) IsSupported() bool {
+	_, err := ks.load()
+	if err != nil {
+		ks.log.Debug("System keyring service is not available", "error", err.Error())
+		return false
+	}
+	return true
+}
+
+// Delete removes a Credential by its guid.
+// If lookup by guid fails, a NotFoundError is returned.
+func (ks *keyringCredentialsService) Delete(guid string) error {
+	table, err := ks.load()
+	if err != nil {
+		return err
+	}
+
+	_, exists := table[guid]
+	if !exists {
+		ks.log.Debug("Credential does not exist", "credential", guid)
+		return NewNotFoundError(guid)
+	}
+
+	delete(table, guid)
+	return ks.save(table)
+}
+
+// Get retrieves a Credential by its guid.
+func (ks *keyringCredentialsService) Get(guid string) (*Credential, error) {
+	table, err := ks.load()
 	if err != nil {
-		if strings.HasSuffix(err.Error(), "exec: \"dbus-launch\": executable file not found in $PATH") {
-			keyring.MockInit()
-			return keyring.Set(service, user, password)
+		return nil, err
+	}
+
+	cr, exists := table[guid]
+	if !exists {
+		ks.log.Debug("Credential does not exist", "credential", guid)
+		return nil, NewNotFoundError(guid)
+	}
+
+	return cr.ToCredential()
+}
+
+// List retrieves all Credentials
+func (ks *keyringCredentialsService) List() ([]Credential, error) {
```

Comment:
Noticed that the `keyring_test.go` file didn't have a test for this function. Maybe handy to add a test ensuring that an empty array is returned instead of `nil` to avoid the bug we have seen in other places: https://github.com/posit-dev/publisher/issues/1937

Reply:
Good catch, pushed a new commit adding a test for it 👍